tests/abstract_model_inheritance/tests.py

```diff
@@ -0,0 +1,306 @@
+from __future__ import unicode_literals
+
+from django.apps import apps
+from django.contrib.contenttypes.fields import (
+    GenericForeignKey, GenericRelation,
+)
+from django.contrib.contenttypes.models import ContentType
+from django.core.checks import Error
+from django.core.exceptions import FieldDoesNotExist, FieldError
+from django.db import models
+from django.test import TestCase
+
+
+class IsolatedModelsTestCase(TestCase):
+
+    def setUp(self):
+        # The unmanaged models need to be removed after the test in order to
+        # prevent bad interactions with the flush operation in other tests.
+        self._old_models = apps.app_configs['abstract_model_inheritance'].models.copy()
+
+    def tearDown(self):
+        apps.app_configs['abstract_model_inheritance'].models = self._old_models
+        apps.all_models['abstract_model_inheritance'] = self._old_models
+        apps.clear_cache()
+
+
+class AbstractInheritanceTests(IsolatedModelsTestCase):
+    def test_single_parent(self):
+        class Parent(models.Model):
+            name = models.CharField(max_length=30)
+
+            class Meta:
+                abstract = True
+
+        class Child(Parent):
+            name = models.CharField(max_length=50)
+
+            class Meta:
+                abstract = True
+
+        class Cousin(Parent):
+            name = models.CharField(max_length=50)
+
+        class GrandChild(Child):
+            pass
+
+        self.assertEqual(Child._meta.get_field('name').max_length, 50)
+        self.assertEqual(Cousin._meta.get_field('name').max_length, 50)
+        self.assertEqual(GrandChild._meta.get_field('name').max_length, 50)
+
+    def test_multiple_parents_mro(self):
+        class GrandMother(models.Model):
+            class Meta:
+                abstract = True
+
+        class GrandFather(models.Model):
+            name = models.CharField(max_length=30)
+
+            class Meta:
+                abstract = True
+
+        class Mother(GrandMother, GrandFather):
+            class Meta:
+                abstract = True
+
+        class Father(GrandMother, GrandFather):
+            name = models.CharField(max_length=50)
+
+            class Meta:
+                abstract = True
+
+        class Child(Mother, Father):
+            pass
+
+        self.assertEqual(Mother._meta.get_field('name').max_length, 30)
+        self.assertEqual(Father._meta.get_field('name').max_length, 50)
+        self.assertEqual(Child._meta.get_field('name').max_length, 50)
+
+    def test_multiple_inheritance_cannot_shadow_concrete_inherited_field(self):
+
+        class ConcreteParent(models.Model):
+            name = models.CharField(max_length=255)
+
+        class AbstractParent(models.Model):
+            name = models.IntegerField()
+
+            class Meta:
+                abstract = True
+
+        class Child(ConcreteParent, AbstractParent):
+            pass
```

Comment:
As far as I tested symmetrical situation is allowed

`class Child(AbstractParent, ConcreteParent):`

It will take `name` field from AbstractParent hiding ConcreteParent required field.